staging/src/k8s.io/api/batch/v1/types.go

```diff
@@ -341,6 +353,22 @@ type JobSpec struct {
 	//
 	// +optional
 	Suspend *bool `json:"suspend,omitempty" protobuf:"varint,10,opt,name=suspend"`
+
+	// podReplacementPolicy specifies when to create replacement Pods.
+	// Possible values are:
+	// - TerminatingOrFailed means to create a replacement Pod when the previously
+	//   created Pod is terminating or failed.
+	// - Failed means to wait until a previously created Pod is fully terminated (has phase
+	//   Failed or Succeeded) before creating a replacement Pod.
+	//
+	// When using podFailurePolicy, a null value for podReplacementPolicy is interpreted
```

Comment:
While I can appreciate that defaulting based on another field can produce unexpected outcomes when podFailurePolicy changes, we have defaulting to avoid client interpretation like this comment describes.  I think the field needs to be conditionally defaulted to take the burden off of clients and a transformation from not specifying podFailurePolicy to specifying podFailurePolicy will also need to acknowledge/set the podReplacementPolicy to Failed.

Reply:
That sounds good.  I'll push up a change adding defaulting logic.

Reply:
Updated. PTAL when you can.

Reply:
I agree that it would be complicated to interpret the nil value given the dependency on another field, which is why generally we prefer not to do it.

However, given that defaulting happens in every read from etcd, I'm afraid that this would cause a write for every Job object in the store during a control plane upgrade.

On the other hand, I'm not sure what happens when we clear the field in the strategy.

Maybe we can compromise on something: interpret nil as Failed, but set the default if podFailurePolicy is in use (this should at least be a smaller subset of jobs in most clusters).

Reply:
> While I can appreciate that defaulting based on another field can produce unexpected outcomes when podFailurePolicy changes

are the outcomes merely weird, or do they actively break things? e.g. if I apply one job with one podFailurePolicy, this gets defaulted, then I try to reapply the job with a different podFailurePolicy or with no podFailurePolicy, will the defaulted podReplacementPolicy fail validation on update?

Reply:
> I'm afraid that this would cause a write for every Job object in the store during a control plane upgrade.

Do you mean reading would trigger an etcd write, or something else? Newly defaulted fields wouldn't hit etcd until the next actual write of the object (update/patch/etc).

More concerning than triggering writes is returning/persisting data for feature-disabled fields, e.g. alpha fields.

Reply:
FWIIW, podFailurePolicy is immutable, so effectively podReplacementPolicy is immutable if podFailurePolicy is set.

Reply:
> Do you mean reading would trigger an etcd write, or something else? Newly defaulted fields wouldn't hit etcd until the next actual write of the object (update/patch/etc).

Yes, that's what I meant. I though we discussed something like this could happen, but if I'm mistaken, then great, this shouldn't be a problem.

> More concerning than triggering writes is returning/persisting data for feature-disabled fields, e.g. alpha fields.

Even if we clear them in the strategy?

Reply:
> > More concerning than triggering writes is returning/persisting data for feature-disabled fields, e.g. alpha fields.
> 
> Even if we clear them in the strategy?

strategy isn't invoked on read, so if you default it, it'll get returned populated to the user, which we don't want if the field is alpha+feature-disabled

Reply:
What if we set the "default" via mutation in the strategy, based on the feature gate?

Reply:
that could work, or if you do want default-on-read behavior, we do have existing defaults that are conditional on feature enablement

Reply:
The failing flow looks like
1. create job without a podFailurePolicy, the podReplacementPolicy is defaulted to TerminatingOrFailed to be compatible with existing behavior (every job has a replacement policy)
2. add a podFailurePolicy, but without a podReplacementPolicy.  The previous behavior (TerminatingOrFailed) cannot be honored because it produces inconsistent results (pods must be Failed in order to know if a new one is needed).

To add the podFailurePolicy, you do have a acknowledge (add) the podReplacementPolicy because it's substantively changing the meaning of your job when you add the podFailurePolicy.  If we take the addition of the podFailurePolicy to implicitly mean "silently change the podReplacementPolicy" (not clear to me that we want to), we might have enough information in field metadata to do that in a strategy (do defaults get assigned to fieldmanagers?).

Letting each client take a guess at what nil means seems likely to result in inconsistent outcomes for various clients.

Reply:
> that could work, or if you do want default-on-read behavior, we do have existing defaults that are conditional on feature enablement

Just so I understand, this would be sufficent?

```
	if utilfeature.DefaultFeatureGate.Enabled(features.JobPodReplacementPolicy) {
		if obj.Spec.PodReplacementPolicy == nil {
			if obj.Spec.PodFailurePolicy != nil {
				obj.Spec.PodReplacementPolicy = podReplacementPolicyPtr(batchv1.Failed)
			} else {
				obj.Spec.PodReplacementPolicy = podReplacementPolicyPtr(batchv1.TerminatingOrFailed)
			}
		}
	}
```

I see that we have some examples of this in StatefulSet.

Reply:
> FWIIW, podFailurePolicy is immutable, so effectively podReplacementPolicy is immutable if podFailurePolicy is set.

is it also immutable inside cronjob? does the cross-field defaulting also happen inside of a cronjob spec? we probably wouldn't want that

Reply:
> is it also immutable inside cronjob? does the cross-field defaulting also happen inside of a cronjob spec? we probably wouldn't want that

I haven't checked the code to see how it was implemented. I wasn't expecting it to impact cronjobs since the job that gets created in the end will get defaulted.

Reply:
Coded to impact only jobs and the defaulting test for cronjob shows no defaulting: https://github.com/kubernetes/kubernetes/blob/00c7c4d4ee5cc7bea6bcc3768463badd8fa2afbe/pkg/apis/batch/v1/defaults_test.go#L425

Reply:
yup, CronJob defaulting it's a separate function. Also CronJob update validation is a separate function.

We are good to use the feature gate then in job defaulting, then :)